### PR TITLE
[Merged by Bors] - fix resource limit usage

### LINKED ISSUE #161: Compute resources and number of executors are not set
### Affected version

nightly

### Current and expected behavior

Configuring the following compute resources
```
      job:
        resources:
          cpu:
            min: 100m
            max: "1"
          memory:
            limit: 1Gi
      driver:
        resources:
          cpu:
            min: "1"
            max: "2"
          memory:
            limit: 2Gi
        volumeMounts:
          - name: script
            mountPath: /stackable/spark/jobs
      executor:
        instances: 5
        resources:
          cpu:
            min: "4"
            max: "6"
          memory:
            limit: 8Gi
        volumeMounts:
          - name: script
            mountPath: /stackable/spark/jobs
```
I would have expected the following settings (i manually overwrite them to get it working)
```
      sparkConf:
        spark.executor.instances: "5"
        spark.driver.cores: "2"
        spark.driver.memory: "2g"
        spark.kubernetes.driver.request.cores: "1"
        spark.kubernetes.driver.limit.cores: "2"
        spark.executor.cores: "6"
        spark.executor.memory: "8g"
        spark.kubernetes.executor.request.cores: "4"
        spark.kubernetes.executor.limit.cores: "6"
```

Whatever currently is done is way of, e.g. it only spawns 2 executors, executor cpu and memory seem to be the default values and the executor thinks it has a single core.

### Possible solution

Did some prototyping in https://github.com/stackabletech/spark-k8s-operator/pull/160
In a nutshell we need to set all the above mentioned values (can be used as a test case)
* `spark.executor.instances` obviously
* `spark.driver.memory` and `spark.executor.memory` need to be fetched from the memory limit and somehow converted to the Spark memory convention. Special caution is needed as if a user specifies e.g. 8Gi for a executor, Spark will add overhead and the resulting Pods will have a memory limit of `11468Mi`. We want to subtract that overhead and do calculations (and/or set overhead configurations) to achieve the requested 8Gi memory limit of the executor Pods.
* spark.kubernetes.driver.(request|limit).cores and spark.kubernetes.driver.(request|limit).cores can be 1:1 mapped from the CRD i guess
* `spark.driver.cores` and `spark.executor.cores` need to be fetched from the cpu limit, somehow converted to a floating point, rounded up and given as a while positive number

### Additional context

_No response_

### Environment

_No response_

### Would you like to work on fixing this bug?

_No response_

### PR DESCRIPTION
**Pull request by adwk67**
# Description

- Restores setting # of executor instances
- Define spark conf settings as well as defining the pod template
  - set kuberenetes.driver/executor.* from CRD directly
  - use cpu limit to set driver/executor.cores
- Integration test to test/illustrate
  - just resource limits
  - just spark.conf settings
  - just defaults
  
Fixes #161.
Jenkins: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/41/

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: tests/templates/kuttl/resources/12-deploy-spark-app.yaml.j2 [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  mode: cluster
+  mainApplicationFile: local:///stackable/spark/examples/src/main/python/streaming/hdfs_wordcount.py
+  args:
+    - "/tmp2"
+  sparkConf:
+    spark.kubernetes.submission.waitAppCompletion: "false"
+    spark.kubernetes.driver.pod.name: "resources-sparkconf-driver"
+    spark.kubernetes.executor.podNamePrefix: "resources-sparkconf"
+    spark.kubernetes.driver.request.cores: "2"
+    spark.kubernetes.driver.limit.cores: "3"
```
Comment (sbernauer) [inline tests/templates/kuttl/resources/12-deploy-spark-app.yaml.j2:18]: As this serves kind of howto i would prefer to include this setting
```suggestion
    spark.kubernetes.driver.limit.cores: "3"
    spark.driver.cores: "3"
```

File: tests/templates/kuttl/resources/12-deploy-spark-app.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    spark.driver.memory: "1g"
+    spark.driver.memoryOverheadFactor: "0.4"
+    spark.kubernetes.executor.request.cores: "2"
+    spark.kubernetes.executor.limit.cores: "3"
```
Comment (sbernauer) [inline tests/templates/kuttl/resources/12-deploy-spark-app.yaml.j2:22]: As this serves kind of howto i would prefer to include this setting
```suggestion
    spark.kubernetes.executor.limit.cores: "3"
    spark.executor.cores: "3"
```

**Review by sbernauer — COMMENTED** [review]
Code LGTM overall
I'm going to give it a try by using it within the warehouse demo

File: rust/crd/src/lib.rs [hunk 3]
```diff
@@ -417,6 +507,32 @@ impl SparkApplication {
         Ok(submit_cmd)
     }
 
+    // a memory overhead will be applied using a factor of 0.1 (JVM jobs) or 0.4 (non-JVM jobs),
+    // being not less than 384MB. The resource limit should keep this transparent by reducing the
+    // declared memory limit accordingly.
+    fn jvm_memory_format(&self, limit: &Quantity) -> Result<String, Error> {
```
[the remark this answers is not in the document]
Comment (adwk67) [reply rust/crd/src/lib.rs]: Agreed - done!

**Comment by sbernauer** [comment]
Tests it out and seemed to work fine :+1:

File: rust/crd/src/lib.rs [hunk 4]
```diff
@@ -398,6 +391,103 @@ impl SparkApplication {
...
+            ..
+        }) = &self.executor_resources()
+        {
+            submit_conf.insert(
+                "--conf spark.kubernetes.executor.limit.cores".to_string(),
+                max.0.clone(),
+            );
+            let cores =
+                cores_from_quantity(max.0.clone()).map_err(|_| Error::FailedQuantityConversion)?;
+            submit_conf.insert("--conf spark.executor.cores".to_string(), cores);
```
[the remark this answers is not in the document]
Comment (adwk67) [reply rust/crd/src/lib.rs:448]: Done.

**Review by sbernauer — APPROVED** [review]
LGTM % remaining two comments. worked fine :+1:

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -398,6 +391,103 @@ impl SparkApplication {
             submit_cmd.extend(deps.packages.map(|p| format!("--packages {}", p.join(","))));
         }
 
+        // some command elements need to be initially stored in a map (to allow overwrites) and
+        // then added to the vector once complete.
+        let mut submit_conf: BTreeMap<String, String> = BTreeMap::new();
```
[the remark this answers is not in the document]
Comment (adwk67) [reply rust/crd/src/lib.rs:396]: Agreed and done :)

File: rust/crd/src/lib.rs [hunk 6]
```diff
@@ -398,6 +391,105 @@ impl SparkApplication {
...
+                    instances.to_string(),
+                );
+            }
+        }
+
+        // conf arguments: these should follow - and thus override - values set from resource limits above
+        if let Some(spark_conf) = self.spec.spark_conf.clone() {
+            for (key, value) in spark_conf {
+                submit_conf.insert(key, value);
+            }
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:486]:
```suggestion
            submit_conf.extend(spark_conf);
```

**Review by sbernauer — APPROVED** [review]

**Comment by adwk67** [comment]
bors merge

**Pull request closed without merging**